### PR TITLE
Remove pluggability of ZenPing

### PR DESCRIPTION
**Pull request by rjernst**
ZenPing is the part of zen discovery which knows how to ping nodes.
There is only one alternative implementation, which is just for testing.
This change removes the ability to add custom zen pings, and instead
hooks in the MockZenPing for tests through an overridden method in
MockNode. This also folds in the ZenPingService (which was really just a
single method) into ZenDiscovery, and removes the idea of having
multiple ZenPing instances. Finally, this was the last usage of the
ExtensionPoint classes, so that is also removed here.

### CODE REVIEW

File: docs/reference/modules/network.asciidoc [hunk 1]
```diff
@@ -26,7 +26,7 @@ of these.
 +
 Defaults to `_local_`.
 
-`discovery.zen.ping.unicast.hosts`::
+`discovery.zen.pinger.unicast.hosts`::
```
Comment (jasontedor) [inline docs/reference/modules/network.asciidoc]: I suspect this was unintentional. I see replacements of `ping` to `pinger` throughout. It should be reverted though.

**Review by jasontedor — CHANGES_REQUESTED** [review]
I can give this a full review after this `ping` -> `pinger` thing is fixed.

**Comment by rjernst** [comment]
Oops sorry, intellij was a little aggressive in my rename. I pushed e855b60.

**Comment by jasontedor** [comment]
@rjernst Can you resolve the failing build, it looks indicative of a real issue?

**Comment by rjernst** [comment]
I pushed b263aff which fixed the test for me. It was a test only issue, for which I had to modify the ZenPing interface a bit.

**Comment by rjernst** [comment]
The remaining test failure is another test issue with TribeIT. The problem is TribeService creates Nodes directly (actually a subclass, TribeClientNode, which does not do Environment initialization). With the zen ping override being moved into MockNode, the TribeClientNodes now know nothing about MockZenPing, and so try starting unicast zen ping.  I have a fix (mocking tribe service to create mock TribeClientNodes and having a MockNode override for creating the tribe service), but am still working through getting it just right. I will update this when it is ready for review again.

**Comment by rjernst** [comment]
Ohhh, actually this turned out to be simpler than I thought. I figured out a way around having TribeClientNode altogether, and allowed delegating to the Node instance itself on constructing an inner client node for tribe. I think this is ready for review again @jasontedor.

File: core/src/main/java/org/elasticsearch/discovery/zen/FaultDetection.java [hunk 2]
```diff
@@ -36,7 +38,7 @@
  * A base class for {@link MasterFaultDetection} &amp; {@link NodesFaultDetection},
  * making sure both use the same setting.
  */
-public abstract class FaultDetection extends AbstractComponent {
+public abstract class FaultDetection extends AbstractComponent implements Closeable {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/zen/FaultDetection.java]: why is this related?

File: core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java [hunk 3]
```diff
@@ -105,7 +107,7 @@
     private AllocationService allocationService;
     private final ClusterName clusterName;
     private final DiscoverySettings discoverySettings;
-    private final ZenPingService pingService;
+    private final ZenPing pinger;
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java]: nit: can we call this variable zenPing ?

File: core/src/main/java/org/elasticsearch/discovery/zen/UnicastZenPing.java [hunk 4]
```diff
@@ -501,9 +485,6 @@ public void handleException(TransportException exp) {
     }
 
     private UnicastPingResponse handlePingRequest(final UnicastPingRequest request) {
-        if (!lifecycle.started()) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/zen/UnicastZenPing.java]: great we can't end up here any more with a non-started contextProvider.

File: core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java [hunk 5]
```diff
@@ -1009,6 +1003,33 @@ public void handleException(TransportException exp) {
         }
     }
 
+    private ZenPing.PingCollection pingAndWait(TimeValue timeout) {
+        final ZenPing.PingCollection response = new ZenPing.PingCollection();
+        final CountDownLatch latch = new CountDownLatch(1);
+        final AtomicBoolean counted = new AtomicBoolean();
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java]: nit: we can double countDown() the latch, so I don't think we need the counted boolean?

File: core/src/main/java/org/elasticsearch/node/Node.java [hunk 6]
```diff
@@ -881,4 +889,17 @@ protected SearchService newSearchService(ClusterService clusterService, IndicesS
         }
         return customNameResolvers;
     }
+
+    /** Create a new ZenPing instance for use in zen discovery. */
+    protected ZenPing newZenPing(Settings settings, ThreadPool threadPool, TransportService transportService,
+                                 UnicastHostsProvider hostsProvider) {
+        Logger logger = Loggers.getLogger(getClass(), settings);
+        logger.info("Creating unicast zen ping");
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/node/Node.java]: I think we can remove this log line as it's normal?

File: core/src/main/java/org/elasticsearch/tribe/TribeService.java [hunk 7]
```diff
@@ -185,15 +185,15 @@ public static Settings processSettings(Settings settings) {
...
+                        Function<Settings, Node> clientNodeBuilder) {
         super(settings);
         this.clusterService = clusterService;
         Map<String, Settings> nodesSettings = new HashMap<>(settings.getGroups("tribe", true));
         nodesSettings.remove("blocks"); // remove prefix settings that don't indicate a client
         nodesSettings.remove("on_conflict"); // remove prefix settings that don't indicate a client
         for (Map.Entry<String, Settings> entry : nodesSettings.entrySet()) {
             Settings clientSettings = buildClientSettings(entry.getKey(), tribeNodeId, settings, entry.getValue());
-            nodes.add(new TribeClientNode(clientSettings, classpathPlugins));
+            nodes.add(clientNodeBuilder.apply(clientSettings));
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/tribe/TribeService.java]: +++

File: qa/evil-tests/src/test/java/org/elasticsearch/tribe/TribeUnitTests.java [hunk 8]
```diff
@@ -69,14 +69,14 @@ public static void createTribes() throws NodeValidationException {
...
         final List<Class<? extends Plugin>> mockPlugins = Arrays.asList(MockTcpTransportPlugin.class, MockZenPing.TestPlugin.class);
-        tribe1 = new TribeClientNode(
+        tribe1 = new MockNode(
             Settings.builder()
                 .put(baseSettings)
                 .put("cluster.name", "tribe1")
                 .put("node.name", "tribe1_node")
                     .put(NodeEnvironment.NODE_ID_SEED_SETTING.getKey(), random().nextLong())
                 .build(), mockPlugins).start();
-        tribe2 = new TribeClientNode(
```
Comment (bleskes) [inline qa/evil-tests/src/test/java/org/elasticsearch/tribe/TribeUnitTests.java]: +1

File: test/framework/src/main/java/org/elasticsearch/node/MockNode.java [hunk 9]
```diff
@@ -97,6 +100,21 @@ protected TransportService newTransportService(Settings settings, Transport tran
     }
 
     @Override
+    protected ZenPing newZenPing(Settings settings, ThreadPool threadPool, TransportService transportService,
+                                 UnicastHostsProvider hostsProvider) {
+        if (getPluginsService().filterPlugins(MockZenPing.TestPlugin.class).isEmpty()) {
```
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/node/MockNode.java]: doesn't have to be part of this change, but I wonder if we should use a test only setting for this. Might be simpler.

File: test/framework/src/main/java/org/elasticsearch/test/discovery/MockZenPing.java [hunk 10]
```diff
@@ -40,7 +41,9 @@
  * A {@link ZenPing} implementation which returns results based on an static in-memory map. This allows pinging
  * to be immediate and can be used to speed up tests.
  */
-public final class MockZenPing extends AbstractLifecycleComponent implements ZenPing {
+public final class MockZenPing extends AbstractComponent implements ZenPing {
+
+    public static class TestPlugin extends Plugin {}
```
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/test/discovery/MockZenPing.java]: if we keep this, can we add a comment about what it does? I think it will confuse people like this.

**Review by bleskes — CHANGES_REQUESTED** [review]
Looks great Ryan. Thanks for picking this up. I left some minor comments and one main suggestion, concerning the plugin interface for this.

File: core/src/main/java/org/elasticsearch/discovery/DiscoveryModule.java [hunk 11]
```diff
@@ -62,16 +57,30 @@ public DiscoveryModule(Settings settings, TransportService transportService, Net
...
-                if (hostProviders.put(entry.getKey(), entry.getValue()) != null) {
-                    throw new IllegalArgumentException("Cannot specify zen hosts provider [" + entry.getKey() + "] twice");
-                }
-            });
+        String discoveryType = DISCOVERY_TYPE_SETTING.get(settings);
+        if (discoveryType.equals("none") == false) {
+            Map<String, Supplier<UnicastHostsProvider>> hostProviders = new HashMap<>();
+            hostProviders.put("zen", () -> Collections::emptyList);
+            for (DiscoveryPlugin plugin : plugins) {
+                plugin.getZenHostsProviders(transportService, networkService).entrySet().forEach(entry -> {
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/discovery/DiscoveryModule.java]: I think long term that might be better, but what I have here follows what we have done with other extension points. Long term I think I might like to still have a selection, but have it be automatically wired to the name of the plugin.

File: core/src/main/java/org/elasticsearch/discovery/zen/FaultDetection.java (hunk 2 above)
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/discovery/zen/FaultDetection.java]: This allowed using `IOUtils.close`

File: core/src/main/java/org/elasticsearch/node/Node.java (hunk 6 above)
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/node/Node.java]: Ooops, this was from debugging. Will remove.

File: test/framework/src/main/java/org/elasticsearch/node/MockNode.java (hunk 9 above)
Comment (rjernst) [inline test/framework/src/main/java/org/elasticsearch/node/MockNode.java]: I followed the pattern we have used for other test only extensions. If we want to change to using test settings instead of marker interfaces that is fine, but we should do it as a separate change.

File: core/src/main/java/org/elasticsearch/discovery/DiscoveryModule.java (hunk 11 above)
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/discovery/DiscoveryModule.java]: Also, part of my reasoning here for not doing a collection of pings is I don't think installing a plugin should automatically do something, I think it should be conscious choice by the user to say "I want to use this thing provided by the plugin I installed".

File: core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java (hunk 3 above)
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java]: Sure, done.

File: test/framework/src/main/java/org/elasticsearch/test/discovery/MockZenPing.java (hunk 10 above)
Comment (rjernst) [inline test/framework/src/main/java/org/elasticsearch/test/discovery/MockZenPing.java]: Added a javadoc comment.

File: core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java (hunk 5 above)
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java]: I removed counted altogether.

**Comment by rjernst** [comment]
@bleskes @jasontedor I pushed an update.

File: core/src/main/java/org/elasticsearch/discovery/DiscoveryModule.java (hunk 11 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/DiscoveryModule.java]: Ok, fair enough for now. We'll have to re-evaluate this for 6.0 anyway.

**Review by bleskes — APPROVED** [review]
LGTM. Thx ryan.
